core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -315,62 +315,59 @@ object SparkSubmit {
+    if (args.isPython) {
+      // In yarn mode for a python app, add pyspark archives to files
+      // that can be distributed with the job
+      if (clusterManager == YARN) {
+        var pyArchives: String = null
+        val pyArchivesEnvOpt = sys.env.get("PYSPARK_ARCHIVES_PATH")
+        if (pyArchivesEnvOpt.isDefined) {
+          pyArchives = pyArchivesEnvOpt.get
+        } else {
+          if (!sys.env.contains("SPARK_HOME")) {
+            printErrorAndExit("SPARK_HOME does not exist for python application in yarn mode.")
           }
-          val py4jFile = new File(pyLibPath, "py4j-0.8.2.1-src.zip")
-          if (!py4jFile.exists()) {
-            printErrorAndExit("py4j-0.8.2.1-src.zip does not exist for python application " +
-              "in yarn mode.")
+          val pythonPath = new ArrayBuffer[String]
+          for (sparkHome <- sys.env.get("SPARK_HOME")) {
+            val pyLibPath = Seq(sparkHome, "python", "lib").mkString(File.separator)
+            val pyArchivesFile = new File(pyLibPath, "pyspark.zip")
+            if (!pyArchivesFile.exists()) {
+              printErrorAndExit("pyspark.zip does not exist for python application in yarn mode.")
+            }
+            val py4jFile = new File(pyLibPath, "py4j-0.8.2.1-src.zip")
+            if (!py4jFile.exists()) {
+              printErrorAndExit("py4j-0.8.2.1-src.zip does not exist for python application " +
+                "in yarn mode.")
+            }
+            pythonPath += pyArchivesFile.getAbsolutePath()
+            pythonPath += py4jFile.getAbsolutePath()
           }
-          pythonPath += pyArchivesFile.getAbsolutePath()
-          pythonPath += py4jFile.getAbsolutePath()
+          pyArchives = pythonPath.mkString(",")
         }
-        pyArchives = pythonPath.mkString(",")
+        args.pyFiles = mergeFileLists(args.pyFiles, pyArchives)
       }
 
-      pyArchives = pyArchives.split(",").map { localPath =>
-        val localURI = Utils.resolveURI(localPath)
-        if (localURI.getScheme != "local") {
-          args.files = mergeFileLists(args.files, localURI.toString)
-          new Path(localPath).getName
+      // If we're running a python app, set the main class to our specific python runner
+      if (deployMode == CLIENT) {
+        if (args.primaryResource == PYSPARK_SHELL) {
+          args.mainClass = "org.apache.spark.api.python.PythonGatewayServer"
         } else {
-          localURI.getPath
+          // If a python file is provided, add it to the child arguments and list of files to
+          // deploy.
+          // Usage: PythonAppRunner <main python file> <extra python files> [app arguments]
+          args.mainClass = "org.apache.spark.deploy.PythonRunner"
+          args.childArgs = ArrayBuffer(args.primaryResource, args.pyFiles) ++ args.childArgs
+          if (clusterManager != YARN) {
+            args.files = mergeFileLists(args.files, args.primaryResource)
+          }
```

Comment:
just so I didn't miss anything, the only functional changes in this block are that you added YARN checks here and in L364, right?